llvm/test/Feature/Repo/repo_global_ctors_optimizer.ll

```diff
@@ -0,0 +1,88 @@
+; This testcase tests the llvm.global_ctors optimization.
+;
+; This test checks that -O3 is able to delete constructors that become empty
+; only after some optimization passes have run, even if repo pruning happens.
+; CHECK-NOT: @_GLOBAL__sub_I_test.cpp
+
+
+; RUN: rm -f %t.db
+; RUN: env REPOFILE=%t.db opt -mtriple="x86_64-pc-linux-gnu-repo" -O3 -S %s -o %t.ll
+; RUN: env REPOFILE=%t.db llc -mtriple="x86_64-pc-linux-gnu-repo" -filetype=obj %t.ll -o %t.o
+; RUN: env REPOFILE=%t.db opt -mtriple="x86_64-pc-linux-gnu-repo" -O3 -S %s | FileCheck %s
+
+target triple = "x86_64-pc-linux-gnu-elf"
+
+%"struct.(anonymous namespace)::ManualMapEntry" = type { i8* }
+
+@_ZN12_GLOBAL__N_112ManualMapSetE = internal global [1 x %"struct.(anonymous namespace)::ManualMapEntry"] zeroinitializer, align 8
```

Comment:
This test case contains *lot* or IR code. Can it be reduced?

Reply:
No, I did try to simplify the test before I commit the patch. The issue will be gone if I remove any instructions.

Reply:
Really? Surely you just need a global ctor that the optimizer can remove? If that's not true then perhaps the original problem description isn't the whole story.

Reply:
> 
> 
> Really? Surely you just need a global ctor that the optimizer can remove? If that's not true then perhaps the original problem description isn't the whole story.

I will try to explain the test.

The test needs to demo that a global ctor can be removed during the first time build but can't be removed during the second time build,  because the repo pruning pass changes the linkage type of a global variable if it is stored in the database.

To test a global ctor that the optimizer can remove, the 'llvm.global_ctors', '_GLOBAL__sub_I_test.cpp', '__cxx_global_var_init()' and '_ZN12_GLOBAL__N_112ManualMapSetE' are needed. To reproduce the issue, '_ZN12_GLOBAL__N_112ManualMapSetE' need to be stored into the database. If it is only called by  __cxx_global_var_init, it will be removed during the optimisation and will not been stored into the database. Therefore, the '_ZN12_GLOBAL__N_112ManualMapSetE' is not only called by '__cxx_global_var_init' but also by 'main' in the given IR test. The '_ZN12_GLOBAL__N_112ManualMapSet' is a fixup of 'main' function and is stored into the database after the first time build. During the second time build, '_ZN12_GLOBAL__N_112ManualMapSetE ' is pruned and its linkage type is changed from internal global to the available_externally. This broke the global variable optimization for llvm.global_ctors.

I am not able to remove the functions in the IR test, but I can simply the function body. I have updated the test following your advice. Please let me know your thought.

Reply:
I think this is a huge improvement. Can we go further? For example, the string "struct.(anonymous namespace)::ManualMapEntry" appears a number of times; can we use something shorter? Can the named sections be removed? (The repo doesn't yet support a ".text.startup" section anyway.) Attributes #1 and #2 are the same: can one be removed?

The premise here is that the simpler the test code, the easier it should be for the next person to understand what's being tested. That's particularly important if the test starts to fail at some point in the future.

Reply:
Thanks Paul, I have simplified all the issues you mentioned above.

Reply:
Better still. I suspect that we can still go further (do we need the "MapEntry" type?). However, I'm happy that this is simple enough to understand now.

Reply:
> 
> do we need the "MapEntry" type?

This type is used by the functions and variable and makes the global ctor optimizer happen.  In addition. the type is quite simple now.

%"MapEntry" = type { i8* }

Anyway, Thanks for quick reply. I will commit the patch.